prow/lib.sh

```diff
@@ -0,0 +1,161 @@
+      # Set artifact dir based on checkout
+      export ARTIFACTS_DIR="${ARTIFACTS_DIR:-${ARTIFACTS}}"
+    fi
+
+    if [ -z "${PULL_PULL_SHA:-}" ]; then
+      export GIT_SHA="${PULL_BASE_SHA}"
+    else
+      export GIT_SHA="${PULL_PULL_SHA}"
+    fi
+
+  else
+    # Use the current commit.
+    GIT_SHA="$(git rev-parse --verify HEAD)"
+    export GIT_SHA
+  fi
+  GIT_BRANCH="$(git rev-parse --abbrev-ref HEAD)"
+  export GIT_BRANCH
+  gcloud auth configure-docker -q
+}
+
+# Download and unpack istio release artifacts.
+function download_untar_istio_release() {
+  local url_path=${1}
+  local tag=${2}
+  local dir=${3:-.}
+  # Download artifacts
+  LINUX_DIST_URL="${url_path}/istio-${tag}-linux.tar.gz"
+
+  wget  -q "${LINUX_DIST_URL}" -P "${dir}"
+  tar -xzf "${dir}/istio-${tag}-linux.tar.gz" -C "${dir}"
+
+  export ISTIOCTL_BIN="${GOPATH}/src/istio.io/istio/istio-${TAG}/bin/istioctl"
+}
+
+# Cleanup e2e resources.
+function cleanup() {
+  if [[ "${CLEAN_CLUSTERS}" == "True" ]]; then
+    unsetup_clusters
+  fi
+  if [[ "${USE_MASON_RESOURCE}" == "True" ]]; then
+    mason_cleanup
+    cat "${FILE_LOG}"
+  fi
+}
+
+# Set up a GKE cluster for testing.
+function setup_e2e_cluster() {
+  WD=$(dirname "$0")
+  WD=$(cd "$WD" || exit; pwd)
+  ROOT=$(dirname "$WD")
+
+  # shellcheck source=prow/mason_lib.sh
+  source "${ROOT}/prow/mason_lib.sh"
+  # shellcheck source=prow/cluster_lib.sh
+  source "${ROOT}/prow/cluster_lib.sh"
+
+  trap cleanup EXIT
+
+  if [[ "${USE_MASON_RESOURCE}" == "True" ]]; then
+    INFO_PATH="$(mktemp /tmp/XXXXX.boskos.info)"
+    FILE_LOG="$(mktemp /tmp/XXXXX.boskos.log)"
+    OWNER=${OWNER:-"e2e"}
+    E2E_ARGS+=("--mason_info=${INFO_PATH}")
+
+    setup_and_export_git_sha
+
+    get_resource "${RESOURCE_TYPE}" "${OWNER}" "${INFO_PATH}" "${FILE_LOG}"
+  else
+    export GIT_SHA="${GIT_SHA:-$TAG}"
+  fi
+  setup_cluster
+}
+
+function clone_cni() {
+  # Clone the CNI repo so the CNI artifacts can be built.
+  if [[ "$PWD" == "${GOPATH}/src/istio.io/istio" ]]; then
+      TMP_DIR=$PWD
+      cd ../ || return
+      git clone -b master "https://github.com/istio/cni.git"
+      cd "${TMP_DIR}" || return
+  fi
+}
+
+function check_kind() {
+  echo "Checking KinD is installed..."
+  if ! kind --help > /dev/null; then
+    echo "Looks like KinD is not installed."
+    exit 1
+  fi
+}
+
+function setup_kind_cluster() {
+  # Installing KinD
+  check_kind
+
+  # Delete any previous e2e KinD cluster
+  echo "Deleting previous KinD cluster with name=e2e-suite"
+  if ! (kind delete cluster --name=e2e-suite) > /dev/null; then
+    echo "No Found existing kind cluster with name e2e-suite. Continue..."
+  fi
+
+  # Create KinD cluster
+  if ! (kind create cluster --name=e2e-suite); then
+    echo "Could not setup KinD environment. Something wrong with KinD setup. Please check your setup and try again."
+    exit 1
+  fi
+  KUBECONFIG="$(kind get kubeconfig-path --name="e2e-suite")"
+  export KUBECONFIG
+}
+
+function cni_run_daemon() {
```

Comment:
This should be in makefile and tests - no reason it can't be done using 'make test' and only be run in prow.

Reply:
re `prow/lib.sh` - the thinking is this file prow/lib.sh will end up in the `pkg` repo (https://github.com/istio/pkg) as a common file for all to use.  I didn't alter this file - its simply a copy from istio/istio and I expect we will just sync up when the file hits the pkg repo.